### ISSUE TITLE
Support wildcard (*) for profile selection [SPR-9585]

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-framework#14219]
**[Sami Nieminen](https://jira.spring.io/secure/ViewProfile.jspa?name=samppa)** opened **[SPR-9585](https://jira.spring.io/browse/SPR-9585?redirect=false)** and commented

When specifying the profiles, which enable certain bean to be registered, it would be handy to allow wildcard \(*) to be used.

For example, we could have a datasource definition which is required by various profiles, that all use the naming convention database-xyz, such as database-oracle, database-sqlserver. Other profiles do not require the datasource bean to be registered.

Therefore, the datasource bean definition now needs this kind of profile-attribute:

```
<bean id="datasource" profile="database-oracle, database-sqlserver">
```

Every time when we add support to some new database, which needs datasource bean, a new profile name needs to be added.

Better would be, if we could define the profile as:

```
<bean id="datasource" profile="database-*">
```

No need to change this when new database profiles are added.

The wildcarded profile selection would then register the datasource bean no matter if
_-Dspring.profiles.active=database-oracle_
or
_-Dspring.profiles.active=database-sqlserver_
is used.

---

3 votes, 4 watchers

**Comment by spring-issuemaster** [comment]
Bulk closing outdated, unresolved issues. Please, reopen if still relevant.

**Comment by GumanPratapFinzly** [comment]
\Running into this exact limitation in production today. We have a growing set of numbered environments (prod1, prod2, prod3, ...) and a similarly open-ended set of non-prod environments (dev, dev2, dev3, test, test2, qa, qa2, ...). We need a controller disabled in all prod* environments and enabled everywhere else.

Today @Profile forces us to either:

Enumerate every non-prod name explicitly (@Profile({"dev","dev2","dev3","test","test2","qa","qa2",...})), which silently breaks (defaults to enabled in prod) whenever a new non-prod environment is added and someone forgets to update this list, or
Enumerate every prod name with negation (@Profile({"!prod1","!prod2","!prod3",...})), which has the same maintenance problem but in the safer direction new prod environments are enabled-by-default in the annotation unless someone remembers to add them.

Either way, @Profile can't express "any profile matching a naming convention," so we end up writing a custom Condition implementation that does prefix matching against Environment.getActiveProfiles() just to avoid hardcoding an ever-growing list in an annotation. That works, but it's boilerplate that a simple wildcard/prefix expression in @Profile (e.g. "prod*" or "!prod*") would eliminate entirely — matching the "starts-with" naming convention that's common in real deployments (numbered environment suffixes, region suffixes, etc.).

Would love to see this revisited — even basic prefix/suffix wildcard support (not full regex) would cover the majority of real-world cases like ours.

**Comment by sbrannen** [comment]
Hi @GumanPratapFinzly,

While native wildcard pattern matching (such as `@Profile("database-*")`) is still not supported directly in the core Spring Framework, Spring Boot supports a feature called _Profile Groups_ (available since Spring Boot 2.4).

Instead of relying on wildcards, you can group related fine-grained profiles under a single logical profile alias in your `application.properties` or `application.yml` file:

#### application.properties

```properties
spring.profiles.group.database[0]=database-oracle
spring.profiles.group.database[1]=database-sqlserver
```
#### application.yml

```yaml
spring:
  profiles:
    group:
      database:
        - "database-oracle"
        - "database-sqlserver"
```
With this configured, you can annotate your beans using `@Profile("database")`, and it will automatically cover the grouped sub-profiles when active.

For more details, see the official [Spring Boot Documentation on Profile Groups](https://docs.spring.io/spring-boot/reference/features/profiles.html#features.profiles.groups).

----

Does that meet your needs?
